lib/matplotlib/backend_bases.py

```diff
@@ -3195,3 +3205,507 @@ def zoom(self, *args):
     def set_history_buttons(self):
         """Enable or disable back/forward button"""
         pass
+
+
+class ToolEvent(object):
+    """Event for tool manipulation (add/remove)"""
+    def __init__(self, name, sender, tool, data=None):
+        self.name = name
+        self.sender = sender
+        self.tool = tool
+        self.data = data
+
+
+class ToolTriggerEvent(ToolEvent):
+    """Event to inform  that a tool has been triggered"""
+    def __init__(self, name, sender, tool, canvasevent=None, data=None):
+        ToolEvent.__init__(self, name, sender, tool, data)
+        self.canvasevent = canvasevent
+
+
+class NavigationMessageEvent(object):
+    """Event carring messages from navigation
+
+    Messages are generaly displayed to the user by the toolbar
+    """
+    def __init__(self, name, sender, message):
+        self.name = name
+        self.sender = sender
+        self.message = message
+
+
+class NavigationBase(object):
+    """Helper class that groups all the user interactions for a FigureManager
+
+    Attributes
+    ----------
+    manager: `FigureManager` instance
+    keypresslock: `LockDraw` to know if the `canvas` key_press_event is
+     locked
+    messagelock: `LockDraw` to know if the message is available to write
+    """
+
+    def __init__(self, manager):
+        self.manager = manager
+        self.canvas = manager.canvas
+
+        self._key_press_handler_id = self.canvas.mpl_connect(
+            'key_press_event', self._key_press)
+
+        self._tools = {}
+        self._keys = {}
+        self._toggled = {}
+        self._callbacks = cbook.CallbackRegistry()
+
+        # to process keypress event
+        self.keypresslock = widgets.LockDraw()
+        self.messagelock = widgets.LockDraw()
+
+    def mpl_connect(self, s, func):
+        """Connect event with string *s* to *func*.
+
+        Parameters
+        -----------
+        s : String
+            Name of the event
+
+            The following events are recognized
+
+            - 'tool_message_event'
+            - 'tool_removed_event'
+            - 'tool_added_event'
+            For every tool added a new event is created
+            - 'tool_trigger_TOOLNAME
+            Where TOOLNAME is the id of the tool.
```

Comment:
Indeed we have the `name` passed inside the event, but I think for the users, it is pretty useful to be able to `connect` to a specific `tool` without having to parse by `name`